microsoft-edge/progressive-web-apps-chromium/how-to/widgets.md

```diff
@@ -0,0 +1,495 @@
+      "title": "Previous",
+      "verb": "previous-song"
+    },
+    {
+      "type": "Action.Execute",
+      "title": "Next",
+      "verb": "next-song"
+    }
+  ],
+  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+  "version": "1.5"
+}
+```
+
+Note the `verb` field in the above JSON template. It will be used when handling widget actions in your service worker code. See [Handle widget actions](#handle-widget-actions).
+
+
+<!-- ====================================================================== -->
+## Access widget instances at runtime
+
+You can access widgets and update them from the PWA service worker code. Accessing widgets at runtime is useful in cases such as:
+
+* [Rendering widgets on installation](#render-widgets-on-installation).
+* [Updating widgets on service worker updates](#update-widgets-on-service-worker-updates).
+* [Handling user actions on widgets](#handle-widget-actions).
+* [Updating widgets periodically or when the app changes](#update-widgets-periodically).
+
+Service workers have access to the `self.widgets` object, and the `widgetclick` event which, together, constitute an API that you use to react to changes and access widgets at runtime.
+
+The following sections provide code examples. For a reference of the API, see the [service worker API reference](#service-worker-api-reference).
+
+#### Render widgets on installation
+
+When a PWA is installed, the widgets that the app defines in its manifest are added to the widgets dashboard but not installed yet. A widget is only installed when the user chooses to add the widget from the dashboard.
+
+When a widget is installed, it is not automatically rendered using the `ms_ac_template` and `data` fields of the widget definition.
+
+To render the widget, listen to the `widgetclick` event in your service worker, and update the widget by using the `widgets.updateByTag` function:
+
+```javascript
+// Listen to the widgetclick event.
+self.addEventListener("widgetclick", event => {
+  // Only listen to the "widget-install" action.
+  if (event.action === "widget-install") {
+    // The widget just got installed, render it using renderWidget.
+    // Pass the event.widget object to the function.
+    event.waitUntil(renderWidget(event.widget));
+  }
+});
+
+async function renderWidget(widget) {
+  // Get the template and data URLs from the widget definition.
+  const templateUrl = widget.definition.msAcTemplate;
+  const dataUrl = widget.definition.data;
+
+  // Fetch the template text and data.
+  const template = await (await fetch(templateUrl)).text();
+  const data = await (await fetch(dataUrl)).text();
+
+  // Render the widget with the template and data.
+  await self.widgets.updateByTag(widget.definition.tag, {template, data});
+}
+```
+
+#### Update widgets on service worker updates
+
+When the service worker code changes in a PWA, the browser detects the change, installs, and then later activates the new service worker.
```

Comment:
```suggestion
When the service worker code changes in a PWA, the browser detects the change, installs the new service worker, and then later activates the new service worker.
```
tech review this change - installs what?  I'm guessing it installs the service worker, not sure.

Reply:
I changed this in the last commit. Let me know what you think.